### ISSUE TITLE
Search performance problems with Alfresco Search Services

### ISSUE THREAD

**Issue by hi-ko** [Alfresco/SearchServices#234]
We are trying to migrate an Alfresco CE system from 5.2 with solr4 to Alfresco 6.1 with Alfresco Search Services (we tried 1.3 and 1.4) but we are facing massive performance problems using Alfresco Search Services / Solr6: Searches running on a similar setup take 3-5 x longer (RAM/storage/CPU) having same solr cache config. The strange thing is that cpu/RAM/storage is more or less idle. So ASS doen't seem to scale with given resources.

Please check also Stack Overflow question [Search performance problems with Alfresco Search Services](https://stackoverflow.com/questions/59246076/search-performance-problems-with-alfresco-search-services)

**Comment by michaelsuzukisagi** [comment]
Hi, are you able to elaborate on what is your setup and how you are conducting the migration?
Have you run 1.3 against 5.2?
What is the given resources?
What is the exact problem? are you not able to index or query?

**Comment by hi-ko** [comment]
@michaelsuzukisagi, thanks for your prompt feedback and sorry for just linking the setup.

No - I just run the queries from the migrated system by reindexing the migrated 6.1 repo.
The virtual hardware has the identical resources:
* 8 cpu cores
* 16 GM RAM (we also tried 32 GB RAM without any change
* storage is running on a ssd san with excellent bonnie++ benchmark 
* solr has ~7,000,000 index docs after reindexing

I took the query from the alfresco log for a simple word search to run that search against Alfresco 5.2 solr and ASS 1.3 and ASS 1.4:
````
URL_PART="/solr/alfresco/afts?wt=json&fl=DBID%2Cscore&rows=100&df=keywords&start=0&locale=*&alternativeDic=DEFAULT_DICTIONARY&fq=%7B%21afts%7DAUTHORITY_FILTER_FROM_JSON&fq=%7B%21afts%7DTENANT_FILTER_FROM_JSON&fq=%7B%21afts%7D%2BTYPE%3A%22cm%3Acontent%22+OR+%2BTYPE%3A%22cm%3Afolder%22&fq=%7B%21afts%7D-TYPE%3A%22cm%3Athumbnail%22+AND+-TYPE%3A%22cm%3AfailedThumbnail%22+AND+-TYPE%3A%22cm%3Arating%22+AND+-TYPE%3A%22st%3Asite%22+AND+-ASPECT%3A%22st%3AsiteContainer%22+AND+-ASPECT%3A%22sys%3Ahidden%22+AND+-cm%3Acreator%3Asystem+AND+-QNAME%3Acomment%5C-*&facet=true&facet.field=%40%7Bhttp%3A%2F%2Fwww.alfresco.org%2Fmodel%2Fcontent%2F1.0%7Dcreator&f.%40%7Bhttp%3A%2F%2Fwww.alfresco.org%2Fmodel%2Fcontent%2F1.0%7Dcreator.facet.limit=100&facet.field=%40%7Bhttp%3A%2F%2Fwww.alfresco.org%2Fmodel%2Fcontent%2F1.0%7Dcontent.mimetype&f.%40%7Bhttp%3A%2F%2Fwww.alfresco.org%2Fmodel%2Fcontent%2F1.0%7Dcontent.mimetype.facet.limit=100&facet.field=%40%7Bhttp%3A%2F%2Fwww.alfresco.org%2Fmodel%2Fcontent%2F1.0%7Dmodifier&f.%40%7Bhttp%3A%2F%2Fwww.alfresco.org%2Fmodel%2Fcontent%2F1.0%7Dmodifier.facet.limit=100&facet.field=%40%7Bhttp%3A%2F%2Fecm4u.de%2Fmodel%2Fmail%2F1.0%7DdistributionDomainList&f.%40%7Bhttp%3A%2F%2Fecm4u.de%2Fmodel%2Fmail%2F1.0%7DdistributionDomainList.facet.limit=100&facet.field=%40%7Bhttp%3A%2F%2Fecm4u.de%2Fmodel%2Fmail%2F1.0%7DdistributionList&f.%40%7Bhttp%3A%2F%2Fecm4u.de%2Fmodel%2Fmail%2F1.0%7DdistributionList.facet.limit=100&facet.field=%40%7Bhttp%3A%2F%2Fecm4u.de%2Fmodel%2Fmail%2F1.0%7DsenderDomain&f.%40%7Bhttp%3A%2F%2Fecm4u.de%2Fmodel%2Fmail%2F1.0%7DsenderDomain.facet.limit=100&facet.query=%7B%21afts%7D%40%7Bhttp%3A%2F%2Fwww.alfresco.org%2Fmodel%2Fcontent%2F1.0%7Dcreated%3A%5BNOW%2FDAY-1DAY+TO+NOW%2FDAY%2B1DAY%5D&facet.query=%7B%21afts%7D%40%7Bhttp%3A%2F%2Fwww.alfresco.org%2Fmodel%2Fcontent%2F1.0%7Dcreated%3A%5BNOW%2FDAY-7DAYS+TO+NOW%2FDAY%2B1DAY%5D&facet.query=%7B%21afts%7D%40%7Bhttp%3A%2F%2Fwww.alfresco.org%2Fmodel%2Fcontent%2F1.0%7Dcreated%3A%5BNOW%2FDAY-1MONTH+TO+NOW%2FDAY%2B1DAY%5D&facet.query=%7B%21afts%7D%40%7Bhttp%3A%2F%2Fwww.alfresco.org%2Fmodel%2Fcontent%2F1.0%7Dcreated%3A%5BNOW%2FDAY-6MONTHS+TO+NOW%2FDAY%2B1DAY%5D&facet.query=%7B%21afts%7D%40%7Bhttp%3A%2F%2Fwww.alfresco.org%2Fmodel%2Fcontent%2F1.0%7Dcreated%3A%5BNOW%2FDAY-1YEAR+TO+NOW%2FDAY%2B1DAY%5D&facet.query=%7B%21afts%7D%40%7Bhttp%3A%2F%2Fwww.alfresco.org%2Fmodel%2Fcontent%2F1.0%7Dcontent.size%3A%5B0+TO+10240%5D&facet.query=%7B%21afts%7D%40%7Bhttp%3A%2F%2Fwww.alfresco.org%2Fmodel%2Fcontent%2F1.0%7Dcontent.size%3A%5B10240+TO+102400%5D&facet.query=%7B%21afts%7D%40%7Bhttp%3A%2F%2Fwww.alfresco.org%2Fmodel%2Fcontent%2F1.0%7Dcontent.size%3A%5B102400+TO+1048576%5D&facet.query=%7B%21afts%7D%40%7Bhttp%3A%2F%2Fwww.alfresco.org%2Fmodel%2Fcontent%2F1.0%7Dcontent.size%3A%5B1048576+TO+16777216%5D&facet.query=%7B%21afts%7D%40%7Bhttp%3A%2F%2Fwww.alfresco.org%2Fmodel%2Fcontent%2F1.0%7Dcontent.size%3A%5B16777216+TO+134217728%5D&facet.query=%7B%21afts%7D%40%7Bhttp%3A%2F%2Fwww.alfresco.org%2Fmodel%2Fcontent%2F1.0%7Dcontent.size%3A%5B134217728+TO+MAX%5D&facet.query=%7B%21afts%7D%40%7Bhttp%3A%2F%2Fwww.alfresco.org%2Fmodel%2Fcontent%2F1.0%7Dmodified%3A%5BNOW%2FDAY-1DAY+TO+NOW%2FDAY%2B1DAY%5D&facet.query=%7B%21afts%7D%40%7Bhttp%3A%2F%2Fwww.alfresco.org%2Fmodel%2Fcontent%2F1.0%7Dmodified%3A%5BNOW%2FDAY-7DAYS+TO+NOW%2FDAY%2B1DAY%5D&facet.query=%7B%21afts%7D%40%7Bhttp%3A%2F%2Fwww.alfresco.org%2Fmodel%2Fcontent%2F1.0%7Dmodified%3A%5BNOW%2FDAY-1MONTH+TO+NOW%2FDAY%2B1DAY%5D&facet.query=%7B%21afts%7D%40%7Bhttp%3A%2F%2Fwww.alfresco.org%2Fmodel%2Fcontent%2F1.0%7Dmodified%3A%5BNOW%2FDAY-6MONTHS+TO+NOW%2FDAY%2B1DAY%5D&facet.query=%7B%21afts%7D%40%7Bhttp%3A%2F%2Fwww.alfresco.org%2Fmodel%2Fcontent%2F1.0%7Dmodified%3A%5BNOW%2FDAY-1YEAR+TO+NOW%2FDAY%2B1DAY%5D&facet.query=%7B%21afts%7D%40%7Bhttp%3A%2F%2Fecm4u.de%2Fmodel%2Fmail%2F1.0%7Dreceived%3A%5BNOW%2FDAY-1DAY+TO+NOW%2FDAY%2B1DAY%5D&facet.query=%7B%21afts%7D%40%7Bhttp%3A%2F%2Fecm4u.de%2Fmodel%2Fmail%2F1.0%7Dreceived%3A%5BNOW%2FDAY-7DAYS+TO+NOW%2FDAY%2B1DAY%5D&facet.query=%7B%21afts%7D%40%7Bhttp%3A%2F%2Fecm4u.de%2Fmodel%2Fmail%2F1.0%7Dreceived%3A%5BNOW%2FDAY-1MONTH+TO+NOW%2FDAY%2B1DAY%5D&facet.query=%7B%21afts%7D%40%7Bhttp%3A%2F%2Fecm4u.de%2Fmodel%2Fmail%2F1.0%7Dreceived%3A%5BNOW%2FDAY-6MONTHS+TO+NOW%2FDAY%2B1DAY%5D&facet.query=%7B%21afts%7D%40%7Bhttp%3A%2F%2Fecm4u.de%2Fmodel%2Fmail%2F1.0%7Dreceived%3A%5BNOW%2FDAY-1YEAR+TO+NOW%2FDAY%2B1DAY%5D&facet.query=%7B%21afts%7D%40%7Bhttp%3A%2F%2Fecm4u.de%2Fmodel%2Fmail%2F1.0%7Dsent%3A%5BNOW%2FDAY-1DAY+TO+NOW%2FDAY%2B1DAY%5D&facet.query=%7B%21afts%7D%40%7Bhttp%3A%2F%2Fecm4u.de%2Fmodel%2Fmail%2F1.0%7Dsent%3A%5BNOW%2FDAY-7DAYS+TO+NOW%2FDAY%2B1DAY%5D&facet.query=%7B%21afts%7D%40%7Bhttp%3A%2F%2Fecm4u.de%2Fmodel%2Fmail%2F1.0%7Dsent%3A%5BNOW%2FDAY-1MONTH+TO+NOW%2FDAY%2B1DAY%5D&facet.query=%7B%21afts%7D%40%7Bhttp%3A%2F%2Fecm4u.de%2Fmodel%2Fmail%2F1.0%7Dsent%3A%5BNOW%2FDAY-6MONTHS+TO+NOW%2FDAY%2B1DAY%5D&facet.query=%7B%21afts%7D%40%7Bhttp%3A%2F%2Fecm4u.de%2Fmodel%2Fmail%2F1.0%7Dsent%3A%5BNOW%2FDAY-1YEAR+TO+NOW%2FDAY%2B1DAY%5D&hl=true&hl.q=$1&hl.snippets=255&hl.fragsize=100&hl.mergeContiguous=false&hl.usePhraseHighlighter=true&hl.simple.pre=%02&hl.simple.post=%03&f.cm%3Aname.hl.snippets=255&f.cm%3Aname.hl.fragsize=100&f.cm%3Aname.hl.fragsize=100&f.cm%3Aname.hl.mergeContiguous=false&f.cm%3Aname.hl.simple.pre=%02&f.cm%3Aname.hl.simple.post=%03&f.cm%3Adescription.hl.snippets=255&f.cm%3Adescription.hl.fragsize=100&f.cm%3Adescription.hl.fragsize=100&f.cm%3Adescription.hl.mergeContiguous=false&f.cm%3Adescription.hl.simple.pre=%02&f.cm%3Adescription.hl.simple.post=%03&f.cm%3Atitle.hl.snippets=255&f.cm%3Atitle.hl.fragsize=100&f.cm%3Atitle.hl.fragsize=100&f.cm%3Atitle.hl.mergeContiguous=false&f.cm%3Atitle.hl.simple.pre=%02&f.cm%3Atitle.hl.simple.post=%03&f.content.hl.snippets=255&f.content.hl.fragsize=100&f.content.hl.fragsize=100&f.content.hl.mergeContiguous=false&f.content.hl.simple.pre=%02&f.content.hl.simple.post=%03&f.ia%3AdescriptionEvent.hl.snippets=255&f.ia%3AdescriptionEvent.hl.fragsize=100&f.ia%3AdescriptionEvent.hl.fragsize=100&f.ia%3AdescriptionEvent.hl.mergeContiguous=false&f.ia%3AdescriptionEvent.hl.simple.pre=%02&f.ia%3AdescriptionEvent.hl.simple.post=%03&f.ia%3AwhatEvent.hl.snippets=255&f.ia%3AwhatEvent.hl.fragsize=100&f.ia%3AwhatEvent.hl.fragsize=100&f.ia%3AwhatEvent.hl.mergeContiguous=false&f.ia%3AwhatEvent.hl.simple.pre=%02&f.ia%3AwhatEvent.hl.simple.post=%03&f.lnk%3Atitle.hl.snippets=255&f.lnk%3Atitle.hl.fragsize=100&f.lnk%3Atitle.hl.fragsize=100&f.lnk%3Atitle.hl.mergeContiguous=false&f.lnk%3Atitle.hl.simple.pre=%02&f.lnk%3Atitle.hl.simple.post=%03&hl.fl=cm%3Aname%2Ccm%3Adescription%2Ccm%3Atitle%2Ccontent%2Cia%3AdescriptionEvent%2Cia%3AwhatEvent%2Clnk%3Atitle&spellcheck.q=$1&spellcheck=true"


JSON_BODY="
{
	\" tenants \": [\" \"],
	\" defaultNamespace \": \" http: //www.alfresco.org/model/content/1.0\",
	\"textAttributes\": [],
	\"locales\": [\"*\"],
	\"defaultFTSOperator\": \"AND\",
	\"defaultFTSFieldOperator\": \"AND\",
	\"anyDenyDenies\": true,
	\"query\": \"$1 \",
	\"templates\": [{
			\"template\": \"%(cm:name cm:title cm:description ia:whatEvent ia:descriptionEvent lnk:title lnk:description TEXT TAG)\",
			\"name\": \"keywords\"
		}
	],
	\"allAttributes\": [],
	\"queryConsistency\": \"DEFAULT\",
	\"authorities\": [\"GROUP_EVERYONE\", \"ROLE_ADMINISTRATOR\", \"ROLE_AUTHENTICATED\", \"ext-robert_h-adm\"]
}
"
curl -k --cert "$REPO_CERT":"$REPO_CERT_PW" \
-H "Content-Type: application/json" -X POST \
-d "$JSON_BODY" \
"$BASE_URL$URL_PART"
````

indexing and search work in principle but the search takes in 5.2 ~1-2 sec but 7-30 sec in ASS.
The query times are even the same when giving much more RAM/heap space (30 GB heap instead of 12 running solr4).
I also tried to increase the default caches without any change in the response times of not cached responses.

I'm wondering why sol4 admin UI shows using 80% of the jvm heap while solr6 doesn't take more than 2-3 GB heap during searches.

**Comment by michaelsuzukisagi** [comment]
Thanks for the details if possible could you let me know:
What is the total memory available to the machine?
Is Alfresco Search Services running on the same machine as repo?
What was Solr4 setup (did it use SAN as well)?
Finally can you provide us the summary report of Alfresco Search services and one for solr4 by calling the following url:
`SOLR_HOST/solr/admin/cores?action=summary&wt=xml`
`SOLR_HOST/solr4/admin/cores?action=summary&wt=xml`

**Comment by hi-ko** [comment]
solr is running on a separate machine.
solr4: 16 GB RAM / jvm started with -mx 12 GB
ASS1.4: 32 GB RAM / jvm started with -mx 30 GB

You can download the summary reports for the alfresco core for both systems [here](https://files.ecm4u.de/s/csWoKpbLsiLxazi)

**Comment by hi-ko** [comment]
Both systems run on the same SAN which is no issue running bonnie++ on both systems having very similar numbers. We tried to have no differences in the setup with the exception that we gave more memory to the ASS setup

**Comment by hi-ko** [comment]
I was able to configure 4 shards (served by only one solr instance) but the performance doesn't change so I guess the bottleneck is the solr/alfresco implementation or the index structure.

**Comment by michaelsuzukisagi** [comment]
Thanks for the update, just to confirm is this in a sharded mode? as per your other ticket (https://github.com/Alfresco/SearchServices/issues/232) if so having all the shards on a single machine is not recommended. Sharding provides a way to manage large index by distributing the size into smaller manageable collections. Having all the shards on a single machine defeats the purpose of having a sharding.

As for the 1.4 setup I would leave more memory for the os and not allocate the majority of the system memory to the JVM running solr.

**Comment by hi-ko** [comment]
yes, this config was just to compare if sharding changes anything in better parallelization independant from the jvm. I would of course prefer avoiding any sharding. There must be a change in implenentation/configuration from alfresco5.2/solr4 to alfresco6/ASS/solr6 which throttles search performance and it would be the worst solution to optimize by multiplating the boxes.

**Comment by michaelsuzukisagi** [comment]
Considering it already has 6,758,195 documents indexed and I assume this would grow, you should consider start sharding the index. Please let me know if you still have issues with the query response time after setting up 1.4 with shard (each shard should be on a separate machine).

**Comment by hi-ko** [comment]
Hi Michael,
I can't belive this should be the soluiton. Using Alfresco5.2/sol4 we had 20 Mio docs in the index in other systems without any problems. Now 5 Mio index docs should be too much? I guess ASS and/or solr6 introduced some serious design flaws. This means we should try to work around design flaws with more hardware.

**Comment by hi-ko** [comment]
as of your official docs: https://docs.alfresco.com/5.2/references/sharding-best-practices.html
Do I need sharding?
If you plan to store 50 million + documents in your repository, you should consider sharding to maximize indexing performance and to enable horizontal scaling to massive content repositories.

**Comment by michaelsuzukisagi** [comment]
It's your choice on how you want to deploy, eventually you will reach a limit. My understanding is you deployed 4 shards on to a single solr instance. This will not work well as stated above. 
If you are still experiencing performance issues after correctly setting up Alfresco Search Services, please raise a new ticket with details of your setup.

**Comment by michaelsuzukisagi** [comment]
Additional information regarding the index https://hub.alfresco.com/t5/alfresco-content-services-blog/sizing-my-alfresco-search-services/ba-p/294052

**Comment by hi-ko** [comment]
I have both configs on the same machine which I switch for testing: 

- no shards / single index
-  4 shards running in 1 jvm.

the prefered configuration is of course to have only one single index. Both configs are more or less the same in terms of search performance but ~ 5 times slower than Alfresco 5.2/solr4 for single word searches.
Do you have any idea why the search performance is so much worse in ASS compared to Alfresco5.2/solr4?

**Comment by michaelsuzukisagi** [comment]
Is the repo still configured to query all the shards even when it's not using shards?
Im still not clear how the performance is the same for a single instance vs 4 shards on a single instance. The latter would have been very slow and non performant.

**Comment by hi-ko** [comment]
No, I commented out the solr6.store.mappings when accessing the single instance unsharded index

**Comment by hi-ko** [comment]
strange thing is that I see only ~2 sec IO the rest of the time spend on cpu.
I will setup an environment having sol4 and ASS 1.4 installed on the same box to exclude any side effects.

**Comment by hi-ko** [comment]
I also compared cache hit ratio which is similar between the ASS-1.4 and the solr4 installation.
I also removed facets and highlighting from the curl query but the ASS search is still ~7 sec compared to 2-3 sec with faceting and highlighting. The most obvious difference is that ASS doesn't use more than ~3 GB of jvm memory.
The main question is: If from the OS point of view the main IO only lasts 2-3 sec, what is the cpu doing the rest of the time ...?
I wouldn't focus on increasing caches since the cache hit ratio is not that good and we can't put the 100 GB index into the cache. Memory seems to be unused anyway once reindex is finished.

**Comment by michaelsuzukisagi** [comment]
Can you provide us with the screen shot for the jvm usage for both solr4 and Alfresco Search Services.

**Comment by hi-ko** [comment]
sure - If you reopen the ticket ;-)

**Comment by hi-ko** [comment]
For the OOM/GC Problem (#233) we found a work around but the performance problem still is unchanged in the latest version 1.4.2
We compare solr4 (separate tomcat) with ASS 1.4.2 (zip install on a separat OS) indexing more or less the same repository (with some minor time gap). Both systems run on dedicated hardware on the same ESX cluster and use the same SAN cluster. IO benchmarks (bonnie+) show similar results on both nodes. 
Both systems use vanilla config (except certificates, paths). We tried to increase caches on the ASS instance but since we didn't see any effect we reverted these changes.

Differences in config:

* solr4
16 GB RAM (12 GB Heap)
oraclejdk8
minimal ubuntu 16.04
* ass 1.4.2
32 GB RAM
adopt openjdk11
minmal ubuntu 18.04

Effect is that the same (uncached) search run from alfresco share takes ~7 sec and on ASS ~40 sec.
We see only a small part of the heap used on ASS while solr4 is using more or less the whole memory and solr4 seams to cache a lot while ass is always reading the whole index (seen with iostat).

Memory/CPU utilisation during a single uncached search:
solr4:
![2020-05-13-15_26_05](https://user-images.githubusercontent.com/5754182/81821187-dd7b4180-9531-11ea-8ee7-79d73e8c7a11.png)
![2020-05-13-15_33_17](https://user-images.githubusercontent.com/5754182/81821196-deac6e80-9531-11ea-8dcb-01e5a976d0cd.png)

ASS 1.4.2
![2020-05-13-15_29_33](https://user-images.githubusercontent.com/5754182/81821333-056aa500-9532-11ea-9521-3fd62de9e548.png)
![2020-05-13-15_33_23](https://user-images.githubusercontent.com/5754182/81821337-06033b80-9532-11ea-9252-88e1df0fcf58.png)

**Comment by aborroy** [comment]
@hi-ko May you re-test this with 1.4.3?

More information available in https://hub.alfresco.com/t5/alfresco-content-services-blog/search-services-1-4-3-release/ba-p/301127

**Comment by hi-ko** [comment]
Great news @aborroy ! I will deploy, reindex and let you know the results ...

**Comment by hi-ko** [comment]
Unfortunately there is no real difference in search performance. Seach with simple words takes >10x longer than on solr4 setup.
The most noticeable difference between solr4 and ass 1.4.x is the low memory usage.
I also see sometimes during search
````
Unexpected thread death: org.eclipse.jetty.util.thread.QueuedThreadPool$2@6906dc7b in qtp2093010349{STARTED,&#8203;10<=10<=10000,&#8203;i=3,&#8203;q=0}
````
Before reindexing I removed all secondary assocs on folders in the system
I compared storage IO on both boxes (tomcat/solr4, ASS 1.4.3) (which use the same SAN) with dd and the throuput is more or less the same (~ 100-200 MB/s).

````time for f in ls /var/lib/solr/index/workspace/SpacesStore/index/*;do dd if="$f" of=/dev/null bs=512 oflag=dsync;done```` 

JVM setup for both systems:

ASS 1.4.3
````
JVM: OpenJDK 64-Bit Server VM (11.0.7+10, mixed mode)
Java: version 11.0.7, vendor AdoptOpenJDK
Java Home: /usr/lib/jvm/adoptopenjdk-11-hotspot-jre-amd64

-Xms16g
-Xmx24g
-XX:+UseConcMarkSweepGC
-XX:-DisableExplicitGC
-XX:NewRatio=3
-XX:SurvivorRatio=4
-XX:TargetSurvivorRatio=90
-XX:MaxTenuringThreshold=8
-XX:ConcGCThreads=4
-XX:ParallelGCThreads=4
-XX:+CMSScavengeBeforeRemark
-XX:PretenureSizeThreshold=64m
-XX:+UseCMSInitiatingOccupancyOnly
-XX:CMSInitiatingOccupancyFraction=50
-XX:CMSMaxAbortablePrecleanTime=6000
-XX:+CMSParallelRemarkEnabled
-XX:+ParallelRefProcEnabled
-XX:-OmitStackTraceInFastThrow
-XX:-UseGCOverheadLimit
-Djetty.home=/opt/alfresco/alfresco-search-services/solr/server
-Dsolr.solr.home=/var/lib/solr/solrhome
-Dsolr.install.dir=/opt/alfresco/alfresco-search-services/solr
-Dlog4j.configuration=file:/var/lib/solr/log4j.properties
-Dsolr.jetty.request.header.size=1000000
-Dsolr.jetty.threads.stop.timeout=300000
-Ddisable.configEdit=true
-Dsolr.allow.unsafe.resourceloading=true
-Djava.net.preferIPv4Stack=true
-Duser.language=en
-Duser.country=US
-Djava.awt.headless=true
-Dfile.encoding=UTF-8
-Djava.net.preferIPv6Addresses=false
-Dsun.security.ssl.allowUnsafeRenegotiation=true
-Xss256k
-Dsolr.log.muteconsole
-XX:OnOutOfMemoryError=/opt/alfresco/alfresco-search-services/solr/bin/oom_solr.sh 8983 /var/log/solr
````

solr4
````
JVM: Java HotSpot(TM) 64-Bit Server VM (25.152-b16, mixed mode)
Java: version 1.8.0_152, vendor Oracle Corporation
Java Home: /usr/lib/jvm/java-8-oracle/jre

-Xms8G
-Xmx12G
-XX:+UseConcMarkSweepGC
-XX:-DisableExplicitGC
-XX:-UseGCOverheadLimit
-Djava.net.preferIPv4Stack=true
-Duser.language=en
-Duser.country=US
-Djava.awt.headless=true
-Dfile.encoding=UTF-8
-Djava.net.preferIPv6Addresses=false
-Dsun.security.ssl.allowUnsafeRenegotiation=true
-Dsolr.allow.unsafe.resourceloading=true
-Dlog4j.configuration=file:///opt/alfresco/solr4/log4j-solr.properties
-Djava.endorsed.dirs=/usr/share/tomcat7/endorsed
-Dcatalina.base=/opt/alfresco/tomcat-solr
-Dcatalina.home=/usr/share/tomcat7
-Djava.io.tmpdir=/var/tmp/tomcat-solr-temp
-Djava.util.logging.config.file=/opt/alfresco/tomcat-solr/conf/logging.properties
-Djava.util.logging.manager=org.apache.juli.ClassLoaderLogManager
````

It looks like solr6 is always reading the whole 85GB index for every search and that takes quite a long time.
On the solr4 system there is a fraction of that IO during search even that system has less memory configured.

**Comment by hi-ko** [comment]
has anybody an explanation why the solr4 setup is using the heap while solr6 fluctuates between 3 and 6 GB heap consumption? I already added `-XX:NewSize=12G` but without difference.

**Comment by hi-ko** [comment]
Finally our search performance issue is identified and solved. The system had patch https://github.com/keensoft/alf-21940-repo installed to address the locale issue introduced with ASS by always sending `locale = '*'` in a search. This seems to be a bad idea.

Removing that patch and reindexing with `alfresco.cross.locale` config set in shared.properties did the job.
Following properties should be _always_ set in `shared.properties` to get expected behavior in search results (which are not set by default). I propose this should be changed in the ASS distribution to avoid complaints from users not finding their content after migrating to Alfresco 6:

````
# Data types that support cross locale/word splitting/token patterns if tokenised
alfresco.cross.locale.datatype.0={http://www.alfresco.org/model/dictionary/1.0}text
alfresco.cross.locale.datatype.1={http://www.alfresco.org/model/dictionary/1.0}content
alfresco.cross.locale.datatype.2={http://www.alfresco.org/model/dictionary/1.0}mltext
````

**Comment by hi-ko** [comment]
Last comment: the use of `alfresco.cross.locale.` parameters does not lead to any noticeable increase in the index size

**Comment by aborroy** [comment]
Thank you very much, @hi-ko 

That was very useful!

I've created a PR in the original project to document this finding:
https://github.com/keensoft/alf-21940-repo/pull/1
